File: agent-os/agent_os/schema.py
"""The agent schema (library spec §2).

Every agent in the library conforms to :class:`AgentSpec`. The registry
(:mod:`agent_os.registry`) enforces this schema at registration time so a
malformed agent can never ship.
"""

from __future__ import annotations

from dataclasses import dataclass, field
from enum import Enum


class Bucket(str, Enum):
    """Functional bucket. A maintenance/prioritization construct for the team;
    owners never see buckets directly (orchestrator can list them on request)."""

    customer_service = "customer_service"
    sales = "sales"
    marketing = "marketing"
    scheduling_ops = "scheduling_ops"
    finance = "finance"
    reputation = "reputation"
    reporting = "reporting"
    system = "system"


class Status(str, Enum):
    existing = "existing"
    new = "new"
    workaround = "workaround"
    stub = "stub"


class Priority(str, Enum):
    P1 = "P1"
    P2 = "P2"
    P3 = "P3"
    P4 = "P4"


class Channel(str, Enum):
    """Output channel. Drives the formatting rules enforced in
    :mod:`agent_os.channels`."""

    sms = "sms"
    email = "email"
    sequence = "sequence"
    report = "report"
    post = "post"
    widget_reply = "widget_reply"
    internal = "internal"


class TriggerType(str, Enum):
    manual = "manual"
    scheduled = "scheduled"
    event_based = "event_based"


# Channels whose output is read by a customer (not just the owner). Bracketed
# placeholders are never acceptable here, and back-channel/internal content must
# never leak into these drafts.
CUSTOMER_FACING_CHANNELS = frozenset(
    {Channel.sms, Channel.email, Channel.post, Channel.widget_reply}
)

# Channels that may contain markdown. Everything not in this set must be plain
# text (no markdown, no asterisks, no headers).
MARKDOWN_CHANNELS = frozenset(
    {Channel.email, Channel.report, Channel.sequence, Channel.internal}
)


@dataclass(frozen=True)
class InputField:
    """A single typed input the agent consumes."""

    name: str
    type: str  # string | number | date | array | object
    required: bool
    description: str
    # Where the orchestrator sources this field from.
    source: str = "owner"  # owner | shared_context


@dataclass(frozen=True)
class SendCaps:
    per_day: int | None = None
    per_month: int | None = None


@dataclass(frozen=True)
class PermissionScope:
    """Permission posture for an agent. Defaults to drafts-only (v1)."""

    default: str = "drafts_only"
    # Hardcoded approval can never be turned off, even after Phase 4 auto-send
    # ships (complaints, quotes, payment escalation).
    require_owner_approval: bool = True
    hardcoded_approval: bool = False
    send_caps: SendCaps = field(default_factory=SendCaps)
    recipient_filter: str = (
        "any"  # any | existing_customers_only | custom_list | completed_service_only
    )
    notes: str = ""


@dataclass(frozen=True)
class Example:
    owner_ask: str
    expected_route: str
    expected_output_excerpt: str


@dataclass(frozen=True)
class AgentSpec:
    """The full, validated specification for one worker agent (library §2)."""

    agent_id: str
    display_name: str
    bucket: Bucket
    status: Status
    build_priority: Priority
    purpose: str
    routes_here_when: tuple[str, ...]
    channel: Channel
    inputs: tuple[InputField, ...]
    shared_context_reads: tuple[str, ...]
    tool_dependencies: tuple[str, ...]
    permission_scope: PermissionScope
    triggers_supported: tuple[TriggerType, ...]
    title_format: str
    body_format: str
    # Names of trace steps the agent SHOULD honestly show. These are declared
    # intent; the runtime trace must still reflect what actually loaded.
    reasoning_trace_steps: tuple[str, ...]
    example_interactions: tuple[Example, ...]
    # Orchestrator routing hints (keyword sets). Not part of the published
    # schema but stored alongside it so routing stays declarative.
    routing_keywords: tuple[str, ...] = ()
    secondary_channels: tuple[Channel, ...] = ()

    # ----- convenience -----
    @property
    def can_use_markdown(self) -> bool:
        return self.channel in MARKDOWN_CHANNELS

    @property
    def is_customer_facing(self) -> bool:
        return self.channel in CUSTOMER_FACING_CHANNELS

File: agent-service/src/agent-os/agents/_schema.ts
/**
 * Agent schema enforcer — the architectural enforcement of the three rules.
 *
 * Every agent is validated against this Zod schema at registry load. Anything
 * that doesn't conform fails CI. The schema extends the §3.2 sketch with two
 * fields the library spec depends on: `never_auto_send` (hardcoded for
 * complaints, quotes, payment escalation) and structured trigger detail (cron
 * expressions + event names) alongside the supported-trigger enums.
 */

import { z } from "zod";
import type { AgentOutput, AgentRunArgs } from "../types/agent.ts";

export const AgentBucket = z.enum([
  "customer_service",
  "sales",
  "marketing",
  "scheduling_ops",
  "finance",
  "reputation",
  "reporting",
  "system",
]);
export type AgentBucket = z.infer<typeof AgentBucket>;

export const AgentChannel = z.enum([
  "sms",
  "email",
  "sequence",
  "report",
  "post",
  "widget_reply",
  "internal",
]);
export type AgentChannel = z.infer<typeof AgentChannel>;

/** Channels that must be plain text (rule 3). */
export const PLAIN_TEXT_CHANNELS: ReadonlySet<AgentChannel> = new Set([
  "sms",
  "post",
  "widget_reply",
]);

export const SharedContextKey = z.enum([
  "business_profile",
  "widget_history",
  "pipeline_state",
  "agent_run_history",
  "kb",
]);

export const AgentSchema = z.object({
  agent_id: z.string().regex(/^[a-z][a-z0-9_]*$/, "agent_id must be snake_case"),
  display_name: z.string().min(1),
  bucket: AgentBucket,
  status: z.enum(["existing", "new", "workaround", "stub"]),
  build_priority: z.enum(["P1", "P2", "P3", "P4"]),
  purpose: z.string().min(1),
  channel: AgentChannel,
  routes_here_when: z.array(z.string()),
  /** Routing signals for the classifier. */
  keywords: z.array(z.string()).default([]),
  strong_signals: z.array(z.string()).default([]),
  shared_context_needed: z.array(SharedContextKey),
  tool_dependencies: z.array(z.string()),
  permission_scope: z.object({
    default: z.literal("drafts_only"), // v1 hardcoded
    require_owner_approval: z.boolean().default(true),
    // Hardcoded: complaint / quote / payment escalation may never auto-send.
    never_auto_send: z.boolean().default(false),
    send_caps: z
      .object({
        per_day: z.number().optional(),
        per_month: z.number().optional(),
        notes: z.array(z.string()).optional(),
      })
      .optional(),
    recipient_filter: z
      .enum([
        "existing_customers_only",
        "any",
        "custom_list",
        "completed_service_only",
        "scheduled_appointments_only",
      ])
      .optional(),
  }),
  triggers_supported: z.array(z.enum(["manual", "scheduled", "event_based"])),
  trigger_detail: z
    .object({
      scheduled_cron: z.array(z.string()).optional(),
      events: z.array(z.string()).optional(),
    })
    .optional(),
  output_format: z.object({
    title_template: z.string(),
    body_constraints: z.object({
      max_length: z.number().optional(),
      no_markdown: z.boolean().default(false),
    }),
  }),
  examples: z
    .array(
      z.object({
        owner_ask: z.string(),
        expected_route: z.string(),
        expected_output_excerpt: z.string(),
      }),
    )
    .min(3, "Every agent must have at least 3 example interactions"),
});

export type AgentSpec = z.infer<typeof AgentSchema>;

/** A registered agent: the validated spec plus its run function. */
export type Agent = AgentSpec & {
  run: (args: AgentRunArgs) => Promise<AgentOutput>;
};

/**
 * Validates a spec and returns the typed Agent. Cross-checks the §0 rules that
 * are statically decidable:
 *  - business_profile is required for every non-internal agent (substrate).
 *  - plain-text channels must declare no_markdown.
 *  - every example must route to this agent.
 */
export function defineAgent(
  spec: unknown,
  run: Agent["run"],
): Agent {
  const parsed = AgentSchema.parse(spec);

  if (parsed.channel !== "internal" && !parsed.shared_context_needed.includes("business_profile")) {
    throw new Error(`[${parsed.agent_id}] must declare business_profile in shared_context_needed`);
  }
  if (PLAIN_TEXT_CHANNELS.has(parsed.channel) && !parsed.output_format.body_constraints.no_markdown) {
    throw new Error(`[${parsed.agent_id}] plain-text channel "${parsed.channel}" must set body_constraints.no_markdown = true`);
  }
  for (const ex of parsed.examples) {
    if (ex.expected_route !== parsed.agent_id) {
      throw new Error(`[${parsed.agent_id}] example expected_route "${ex.expected_route}" must equal agent_id`);
    }
  }

  return { ...parsed, run };
}
